images/converter/default.go

```diff
@@ -60,41 +59,141 @@ type ConvertHooks struct {
 func IndexConvertFuncWithHook(layerConvertFunc ConvertFunc, docker2oci bool, platformMC platforms.MatchComparer, hooks ConvertHooks) ConvertFunc {
-	c := &defaultConverter{
+	return newDefaultWithOpts(convertOpts{
 		layerConvertFunc: layerConvertFunc,
 		docker2oci:       docker2oci,
 		platformMC:       platformMC,
-		diffIDMap:        make(map[digest.Digest]digest.Digest),
 		hooks:            hooks,
+	}).Convert
+}
+
+func NewDefault(opts ...Opt) (*DefaultConverter, error) {
+	var copts convertOpts
+	for _, o := range opts {
+		if err := o(&copts); err != nil {
+			return nil, err
+		}
 	}
-	return c.convert
+
+	return newDefaultWithOpts(copts), nil
+}
+
+func newDefaultWithOpts(copts convertOpts) *DefaultConverter {
+	if copts.platformMC == nil {
+		copts.platformMC = platforms.All
+	}
+
+	return &DefaultConverter{
+		convertOpts: copts,
+	}
+}
+
+type DefaultConverter struct {
+	convertOpts
+}
+
+type diffIDMap struct {
+	diffIDs map[digest.Digest]digest.Digest // key: old diffID, value: new diffID
+	mutex   sync.RWMutex
 }
 
-type defaultConverter struct {
-	layerConvertFunc ConvertFunc
-	docker2oci       bool
-	platformMC       platforms.MatchComparer
-	diffIDMap        map[digest.Digest]digest.Digest // key: old diffID, value: new diffID
-	diffIDMapMu      sync.RWMutex
-	hooks            ConvertHooks
+type leaseMap struct {
+	leaseIDs map[string][]string
+	mutex    sync.Mutex
+	lm       leases.Manager
+}
+
+func (l *leaseMap) Add(ctx context.Context, key string) {
+	leaseID, ok := leases.FromContext(ctx)
+	if ok {
+		l.mutex.Lock()
+		defer l.mutex.Unlock()
+		l.leaseIDs[key] = append(l.leaseIDs[key], leaseID)
+	}
+}
+
+func (l *leaseMap) Clean(key string) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	delete(l.leaseIDs, key)
+}
+
+func (l *leaseMap) Finalize(ctx context.Context, key string, desc *ocispec.Descriptor) error {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	defer delete(l.leaseIDs, key)
+	for _, leaseID := range l.leaseIDs[key] {
+		if err := l.lm.AddResource(ctx, leases.Lease{ID: leaseID}, leases.Resource{
+			ID:   desc.Digest.String(),
+			Type: "content",
+		}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func makeRefKey(desc ocispec.Descriptor) (string, error) {
+	key, err := json.Marshal(desc)
+	if err != nil {
+		return "", err
+	}
+	return string(key), nil
+}
+
+// singleflightDo makes sure that only one conversion execution is
+// in-flight for a given descriptor at a time.
+func (c *DefaultConverter) singleflightDo(ctx context.Context, cs content.Store, desc ocispec.Descriptor, convert ConvertFunc) (*ocispec.Descriptor, error) {
+	if c.sg == nil {
+		return convert(ctx, cs, desc)
+	}
+
+	key, err := makeRefKey(desc)
+	if err != nil {
+		return nil, err
+	}
+
+	c.leases.Add(ctx, key)
+	newDesc, err, _ := c.sg.Do(key, func() (interface{}, error) {
+		return convert(ctx, cs, desc)
+	})
+	if err != nil {
+		c.leases.Clean(key)
```

Comment:
Can we have it also for non-nil case? otherwise `leaseMap.leaseIDs` seems to become bigger and bigger without cleanup during the converter is used.

Reply:
The non-nil case will be clean in `Finalize`, this will reduce the number of lock acquisition by one.